File: src/test/java/at/ac/uibk/dps/cirrina/runtime/online/PingPongTest.java
package at.ac.uibk.dps.cirrina.runtime.online;

import at.ac.uibk.dps.cirrina.classes.collaborativestatemachine.CollaborativeStateMachineClass;
import at.ac.uibk.dps.cirrina.classes.collaborativestatemachine.CollaborativeStateMachineClassBuilder;
import at.ac.uibk.dps.cirrina.data.DefaultDescriptions;
import at.ac.uibk.dps.cirrina.execution.object.context.NatsContext;
import at.ac.uibk.dps.cirrina.execution.object.event.NatsEventHandler;
import at.ac.uibk.dps.cirrina.io.description.CsmlParser;
import at.ac.uibk.dps.cirrina.runtime.OnlineRuntime;
import io.opentelemetry.sdk.autoconfigure.AutoConfiguredOpenTelemetrySdk;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;

public class PingPongTest {

  private static CollaborativeStateMachineClass collaborativeStateMachineClass;

  @BeforeAll
  public static void setUp() {
    var json = DefaultDescriptions.pingPong;

    Assertions.assertDoesNotThrow(() -> {
      collaborativeStateMachineClass = CollaborativeStateMachineClassBuilder.from(
        CsmlParser.parse(json)
      ).build();
    });
  }

  @Disabled
  @Test
  void testPingPongExecute() {
    final var natsServerURL = System.getenv("NATS_SERVER_URL");
    final var natsBucketName = System.getenv("NATS_BUCKET_NAME");
    final var zooKeeperConnectString = System.getenv("ZOOKEEPER_CONNECT_STRING");

    Assumptions.assumeFalse(natsServerURL == null, "Skipping online runtime test");
    Assumptions.assumeFalse(natsBucketName == null, "Skipping online runtime test");
    Assumptions.assumeFalse(zooKeeperConnectString == null, "Skipping online runtime test");

    Assertions.assertDoesNotThrow(() -> {
      final var openTelemetry = AutoConfiguredOpenTelemetrySdk.initialize().getOpenTelemetrySdk();

      try (final var eventHandler = new NatsEventHandler(natsServerURL)) {
        try (final var persistentContext = new NatsContext(true, natsServerURL, natsBucketName)) {
          final var runtime = new OnlineRuntime(
            "runtime",
            eventHandler,
            persistentContext,
            openTelemetry
          );
          //runtime.run();
        }
      }
    });
  }
}

File: src/test/kotlin/at/ac/uibk/dps/cirrina/PingPongTest.kt
package at.ac.uibk.dps.cirrina

import EventHandlerInMemory
import at.ac.uibk.dps.cirrina.data.DefaultDescriptions
import at.ac.uibk.dps.cirrina.di.DaggerTestComponent
import at.ac.uibk.dps.cirrina.di.TestModule
import at.ac.uibk.dps.cirrina.execution.provider.ContextInMemory
import java.time.Duration
import kotlin.time.measureTime
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertDoesNotThrow
import org.junit.jupiter.api.assertTimeout

class PingPongTest {

  @Test
  fun testPingPongExecute() {
    assertTimeout(Duration.ofSeconds(10)) {
      assertDoesNotThrow {
        val eventHandler = EventHandlerInMemory()
        val context = ContextInMemory()

        val runtime =
          DaggerTestComponent.builder()
            .testModule(TestModule(eventHandler, context, DefaultDescriptions.pingPong))
            .build()
            .runtime()

        val duration = measureTime { runtime.run() }
        println("pingPong execution: $duration")

        assertEquals(100_000, context.get("v"))
      }
    }
  }
}
